### ISSUE TITLE
[Feature request] Add support for Decimal type

### ISSUE THREAD

**Issue by Munksgaard** [elixir-explorer/explorer#867]
As per [this comment](https://elixirforum.com/t/adding-support-for-decimal-in-explorer/61855/3?u=munksgaard) I am opening this issue.

I am working with monetary data, for which floats are largely discouraged. As a result, I am using the fairly popular [Decimal](https://hex.pm/packages/decimal) package and the associated Ecto support to store monetary values in my database and process them in my app.

I’m also a big fan of Livebook and Explorer to process my data. However, Explorer does not currently have support for Decimals, which means that I have to manually convert to floats, losing all the benefits of working with Decimal in the first place.

Therefore, I am wondering what it would take to add support for Decimals to Explorer and humbly requesting it for consideration.

**Comment by billylanchantin** [comment]
(Carrying over some of my comment as well)

Polars does support a decimal datatype:

* https://github.com/pola-rs/polars/blob/849953b097e0d15fe107f3683f003228103b425d/py-polars/src/datatypes.rs#L34

If we want to support it as well, the tricky part will be representing things in Elixir. I see two options:

* Add [`Decimal`](https://hexdocs.pm/decimal/readme.html) as a dependency
* Custom datatype like [`Duration`](https://github.com/elixir-explorer/explorer/blob/main/lib/explorer/duration.ex)

I haven't played with the Decimal datatype in Polars or Arrow, so I don't know if it maps nicely to the `Decimal` library's representation. But I have no objections to exploring either option (or some other thing I didn't think of).

**Comment by josevalim** [comment]
There is a reference here: https://arrow.apache.org/docs/r/reference/data-type.html

The implementation is straight-forward, it is stored as a large integer, which we can easily convert to Decimal using [`Decimal.new/3`](https://hexdocs.pm/decimal/Decimal.html#new/3). If a list of Decimals is given as input, we will need to compute the scale, using Decimal.scale. And for precision, we can use the maximum one.

**Comment by billylanchantin** [comment]
👍 I think I'm comfortable adding Decimal as a dep.

@Munksgaard PRs welcome! If you're not able to, then I can take a crack when I have some time.

**Comment by josevalim** [comment]
Yes, we should raise when encoding those. We will already have to do a prepass converting all decimals into int::128 or int::256, so we can raise on NaN and Infinity.

**Comment by Munksgaard** [comment]
> @Munksgaard PRs welcome! If you're not able to, then I can take a crack when I have some time.

I would love to help out, but realistically I won't be able to contribute anything significant in the next couple of weeks :-(

**Comment by billylanchantin** [comment]
Rustler now supports 128 bit integers (thanks for initiating that, @Munksgaard!):

* https://github.com/rusterlium/rustler/pull/600

I think we can consider this issue blocked until rustler does another release. Otherwise we'd have to go off their master branch, which probably isn't worth the headache since they seem to release regularly.

**Comment by LostKobrakai** [comment]
Seems like this has already been released with rustler 0.32.1

**Comment by philss** [comment]
The update of Rustler to `v0.32.1` has landed on main :)

**Comment by alexpearce** [comment]
I had a crack at adding Decimal support and managed to add reading without too much trouble: https://github.com/alexpearce/explorer/commit/709aa6768f4944a07126eb7ae595f449b5153346

I found adding write support more challenging. It seems that Decimal support in polars-core isn't quite as mature as for other dtypes so write support requires a bit more work.

**Comment by billylanchantin** [comment]
@alexpearce Feel free to open a PR. If you're still having trouble, we can discuss there why write support is more difficult.

**Comment by billylanchantin** [comment]
Update on this:

There is decimal support as of #987 (plus some follow on PRs). However we've noticed some inconsistencies which come from Polars. We documented them with #1033.

I think it's fair to say that this issue can be closed, with the caveat that we've marked the feature "unstable" to align with Polars.

Thoughts?

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
